src/briefcase/integrations/virtual_environment.py

```diff
@@ -0,0 +1,255 @@
+import os
+import shutil
+import sys
+from pathlib import Path
+
+from briefcase.config import AppConfig
+from briefcase.console import Console
+from briefcase.exceptions import BriefcaseCommandError
+from briefcase.integrations import subprocess
+from briefcase.integrations.subprocess import SubprocessArgsT
+
+
+class VenvContext:
+    """Context for running commands in a virtual environment.
+
+    Wraps subprocess functions to ensure commands are run in the venv.
+    """
+
+    def __init__(self, tools, venv_path: Path):
+        self.tools = tools
+        self.venv_path = venv_path
+
+    @property
+    def bin_dir(self) -> Path:
+        r"""Return the path to the virtual environment's binary directory.
+
+        :return: <venv>/bin (or <venv>\Scripts on Windows)
+        """
+        return self.venv_path / ("Scripts" if os.name == "nt" else "bin")
+
+    @property
+    def executable(self) -> str:
+        """Path to the Python executable in the virtual environment.
+
+        :return: Absolute filesystem path to the venv's Python executable.
+        """
+        python = self.bin_dir / ("python.exe" if os.name == "nt" else "python")
+        return os.fspath(python)
+
+    @property
+    def env(self) -> dict:
+        """Base environment variables for invoking processes in the venv.
+
+        The returned environment includes the venv's bin directory at the front of the
+        PATH, and sets VIRTUAL_ENV to the venv path. On Windows, PYTHONHOME is removed.
+        """
+
+        env = dict(os.environ)
+        old_path = env.get("PATH", "")
+        env["PATH"] = os.fspath(self.bin_dir) + (
+            os.pathsep + old_path if old_path else ""
+        )
+        env["VIRTUAL_ENV"] = os.fspath(self.venv_path)
+        if os.name == "nt":
+            env.pop("PYTHONHOME", None)
+        return env
+
+    def exists(self) -> bool:
+        """Check whether the virtual environment exists.
+
+        :return True if venv exists and contains a pyvenv.cfg file. else False
+        """
+        return self.venv_path.exists() and (self.venv_path / "pyvenv.cfg").exists()
+
+    def create(self) -> None:
+        """Create the virtual environment(if it doesn't already exist).
+
+        :raises BriefcaseCommandError if creation fails.
+        """
+        if not self.exists():
+            with self.console.wait_bar(
```

Comment:
I'd suggest the console output should be in the calling method, not the venv; that allows the usage to determine how it is rendered.

That likely also means the `if not self.exists()` logic is either not required, or is done as a "raise error if exists" pass.